### PR TITLE
Kalman filter with inputs

### PR DESCRIPTION
**Pull request by mschauer**
@cstjean This would be my design for #4 . Can you give it a try? See the test file `testinput.jl` how to use the new set-up, a varying `b` would be achieved with the `GenericLinearObserver`.
Note that the time runs from `1:n` in the new parts and the prior is assumed to be valid for time `1`. 

Smoother and all convenience wrapper as well as iterators are still missing.

### CODE REVIEW

File: src/Kalman.jl [hunk 1]
```diff
@@ -166,7 +191,7 @@ Kalman filter
 function kalmanfilter(yy::AbstractMatrix, M::LinearHomogSystem) 
     d = size(M.Phi, 1)
     n = size(yy, 2)
-    kalmanfilter!(0:n, yy, zeros(d, n), zeros(d, d, n), zeros(d, d, n), M)
+    kalmanfilter!(1:n, yy, zeros(d, n), zeros(d, d, n), zeros(d, d, n), M)
```
Comment (mschauer) [inline src/Kalman.jl:194]: Ups, this is a mistake while rebasing.

**Comment by cstjean** [comment]
I won't have time to look into it until next week, unfortunately, but so far it looks very nice. I assume that we'll need a function like `observe!` for the transition model?

EDIT: Nevermind, that's your `predict!` function.

**Comment by cstjean** [comment]
It works! I get the same results as with my PR.

> the prior is assumed to be valid for time 1.

If I'm understanding things correctly, you do things this way because it's awkward to talk about the transition from t=0 to t=1, so you'd rather let the user directly specify the post-transition probability? The output `xxf2` now contains the hidden state from `t=2` to `t=n+1`, right?

I'm fine with merging this instead of my PR.

File: src/input.jl [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    H::TH # dxd
+    R::TR # d
+end
+
+
+function predict!(s, x, P, t, u, M) 
+    if s == t # if `s == t`` no time has passed
+        x + u, P, one(M.Phi)
+    else
+        M.Phi*x + M.b + u, M.Phi*P*M.Phi' + M.Q, M.Phi
```
Comment (cstjean) [inline src/input.jl:33]: I think I understand why you have both `M.b` and `u`, but the redundancy makes me a bit uncomfortable. They're the same quantity. Perhaps we should assert that `uu==nothing` when using `LinearEvolution`?

**Review by cstjean — APPROVED** [review]

File: src/input.jl (hunk 2 above)
Comment (mschauer) [reply src/input.jl:33]: How I think of them is that `b` is the intrinsic drift and `u` is an control/external signal.

Comment (cstjean) [reply src/input.jl:33]: Perhaps, but unless I'm mistaken, `u` is unspecified (implicitly 0?) when `GenericLinearEvolution` is used.

Comment (mschauer) [reply src/input.jl:33]: GenericLinearEvolution will dispatch on function https://github.com/mschauer/Kalman.jl/pull/5/files/0826886fe9698002c68cc8c16c3c8c39a6e8d92e#diff-d0bf3c72d95233c17cc7933aeaa30298R37

Comment (cstjean) [reply src/input.jl:33]: Right. What I mean is that with `LinearObservation` (line 33), the evolution is `Phi*x + M.b + u` (transition matrix + intrinsic drift + control/external signal), whereas with `GenericLinearEvolution` (line 39), it's `Phi*x + b`

Anyway, it's no big deal; just seemed a bit inconsistent.

File: src/input.jl [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+Observe `y = Hx + v` where ``v ~ N(0, R)``
+"""
+struct LinearObservation{TH, TR} <: AbstractObservation
+    H::TH # dxd
+    R::TR # d
+end
+
+
+function predict!(s, x, P, t, u, M) 
+    if s == t # if `s == t`` no time has passed
```
Comment (cstjean) [inline src/input.jl:30]: Shouldn't this special-casing also be done in `predict!(s, x, P, t, U, M::GenericLinearEvolution)`?

**Comment by cstjean** [comment]
Perhaps timesteps should be assumed homogeneous, unless using `GenericLinearEvolutionWithInhomogeneousTime()`, in which case they are to be passed in `U`?

**Comment by mschauer** [comment]
In https://github.com/mschauer/Kalman.jl/pull/5/files#diff-d0bf3c72d95233c17cc7933aeaa30298R38
one would set `Phi = I` and `b = 0` to achieve the same effect.

File: src/input.jl [hunk 4]
```diff
@@ -0,0 +1,96 @@
+
+struct GenericLinearObservation <: AbstractObservation
+end
+struct GenericLinearEvolution{TPhi,Tb,TQ} <: AbstractEvolution
```
Comment (cstjean) [inline src/input.jl:4]: Is there a need for this type to be parametric? `GenericLinearObservation` is parameter-less.

Comment (mschauer) [reply src/input.jl:4]: Good catch.

**Comment by cstjean** [comment]
I started building code on top of this PR. Is there any significant reason not to merge it? Small issues can be fixed afterwards...

**Comment by mschauer** [comment]
You are right, *das Bessere ist der Feind des Guten* ("The best is the enemy of the good.")

**Comment by cstjean** [comment]
Doch!

**Pull request merged**